### PR TITLE
[Helm Bugs] Metadata self-repair does not work with provenance file

### PR DESCRIPTION
**Pull request by proteus1121**
https://issues.sonatype.org/browse/NEXUS-21793

### CODE REVIEW

File: nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/restore/HelmRestoreFacetImpl.java [hunk 1]
```diff
@@ -66,10 +67,15 @@ protected void doInit(final Configuration configuration) throws Exception {
   @TransactionalTouchBlob
   public void restore(final AssetBlob assetBlob, final String path) throws IOException {
     StorageTx tx = UnitOfWork.currentTx();
-    HelmAttributes attributes = helmAttributeParser.getAttributesFromInputStream(assetBlob.getBlob().getInputStream());
-    Asset asset = componentRequired(path)
-        ? helmFacet.findOrCreateAsset(tx, path, AssetKind.HELM_PACKAGE, attributes)
-        : helmFacet.findOrCreateAsset(tx, path, AssetKind.HELM_INDEX, attributes);
```
Comment (chernikovmaksym) [inline nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/restore/HelmRestoreFacetImpl.java]: Missing `HELM_INDEX` ?

Comment (proteus1121) [reply nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/restore/HelmRestoreFacetImpl.java]: HelmAttributes should be empty when HELM_INDEX

File: nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java [hunk 2]
```diff
@@ -86,21 +84,9 @@ public UploadResponse handle(Repository repository, ComponentUpload upload) thro
...
-      } else if (fileName.endsWith(TGZ_EXTENSION)) {
-        attributesFromInputStream = helmPackageParser.getAttributesFromInputStream(tempBlob.get());
-        assetKind = AssetKind.HELM_PACKAGE;
-        extension = TGZ_EXTENSION;
-      } else {
-        throw new IllegalArgumentException("Unsupported extension. Extension must be .tgz or .tgz.prov");
-      }
+      AssetKind assetKind = AssetKind.getAssetKindByFileName(fileName);
+      HelmAttributes attributesFromInputStream = helmPackageParser.getAttributes(assetKind, tempBlob.get());
+      String extension = assetKind.getExtension();
```
Comment (mlukaretkyi) [inline nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: You forgot validation for an extension.  
Unsupported extension. Extension must be .tgz or .tgz.prov

Comment (proteus1121) [reply nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: Done

File: nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/AssetKind.java [hunk 3]
```diff
@@ -25,18 +25,35 @@
...
+  HELM_INDEX(CacheControllerHolder.METADATA, null),
+  HELM_PROVENANCE(CacheControllerHolder.CONTENT, ".tgz.prov"),
+  HELM_PACKAGE(CacheControllerHolder.CONTENT, ".tgz");
 
   private final CacheType cacheType;
+  private final String extension;
 
-  AssetKind(final CacheType cacheType) {
+  AssetKind(final CacheType cacheType, final String extenstion) {
     this.cacheType = cacheType;
```
Comment (mlukaretkyi) [inline nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/AssetKind.java]: minor: extenstion -> extension

Comment (proteus1121) [reply nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/AssetKind.java]: Done

File: nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/AssetKind.java [hunk 4]
```diff
@@ -25,18 +25,35 @@
...
+    return AssetKind.HELM_INDEX;
   }
 
   @Nonnull
   public CacheType getCacheType() {
     return cacheType;
   }
+
+  @Nonnull
+  public String getExtension() {
```
Comment (mlukaretkyi) [inline nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/AssetKind.java]: Can be null for HELM_INDEX
Change to Nullable

Comment (proteus1121) [reply nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/AssetKind.java]: I think better define ".yaml" for HELM_INDEX

File: nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java [hunk 5]
```diff
@@ -84,24 +82,15 @@ public UploadResponse handle(Repository repository, ComponentUpload upload) thro
...
-        assetKind = AssetKind.HELM_PROVENANCE;
-        extension = PROVENANCE_EXTENSION;
-      } else if (fileName.endsWith(TGZ_EXTENSION)) {
-        attributesFromInputStream = helmPackageParser.getAttributesFromInputStream(tempBlob.get());
-        assetKind = AssetKind.HELM_PACKAGE;
-        extension = TGZ_EXTENSION;
-      } else {
-        throw new IllegalArgumentException("Unsupported extension. Extension must be .tgz or .tgz.prov");
-      }
+    if (assetKind != AssetKind.HELM_PROVENANCE && assetKind != AssetKind.HELM_PACKAGE) {
```
Comment (chernikovmaksym) [inline nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: && -> ||

File: nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java [hunk 6]
```diff
@@ -84,24 +82,15 @@ public UploadResponse handle(Repository repository, ComponentUpload upload) thro
...
+      throw new IllegalArgumentException("Unsupported extension. Extension must be .tgz or .tgz.prov");
```
Comment (chernikovmaksym) [inline nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: Asset kind is set in recipe by verifying extension. You don't check extension here, only asset kind so message should be something like: "Unsupported asset kind".

File: nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/AssetKind.java [hunk 7]
```diff
@@ -25,18 +25,35 @@
...
-  HELM_PACKAGE(CacheControllerHolder.CONTENT);
+  HELM_INDEX(CacheControllerHolder.METADATA, ".yaml"),
+  HELM_PROVENANCE(CacheControllerHolder.CONTENT, ".tgz.prov"),
+  HELM_PACKAGE(CacheControllerHolder.CONTENT, ".tgz");
 
   private final CacheType cacheType;
+  private final String extension;
 
-  AssetKind(final CacheType cacheType) {
+  AssetKind(final CacheType cacheType, final String extension) {
```
Comment (chernikovmaksym) [inline nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/AssetKind.java]: Question: May it happen that helm asset kind has more than one extension type? Maybe it's better to make a `Set` for extensions?

Comment (chernikovmaksym) [reply nexus-repository-helm/src/main/java/org/sonatype/repository/helm/internal/AssetKind.java]: Discussed

File: nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java (hunk 6 above)
Comment (proteus1121) [reply nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: Done

Comment (chernikovmaksym) [reply nexus-repository-helm/src/main/java/org/sonatype/repository/helm/HelmUploadHandler.java]: Wrong place. It's upload UI handler, not recipe handler. Reverting...

**Pull request merged**